runtime/src/tests.rs

```diff
@@ -100,3 +168,204 @@ fn pallet_poe_availability() {
+                    == Some(AccountId32::new(
+                        testsfixtures::SAMPLE_USERS[AURA_AUTHOR_ID as usize]
+                            .raw_account
+                            .into()
+                    ))
+            );
+        });
+    }
+
+    // Check that Authorship calls back on ImOnline
+    #[test]
+    fn authorship_notifies_imonline() {
+        new_test_ext().execute_with(|| {
+            let block_number = 1;
+            let slot = Slot::from(SLOT_ID);
+            let pre_digest = Digest {
+                logs: vec![DigestItem::PreRuntime(AURA_ENGINE_ID, slot.encode())],
+            };
+            System::reset_events();
+            System::initialize(&block_number, &Default::default(), &pre_digest);
+
+            assert!(!ImOnline::is_online(AURA_AUTHOR_ID));
+            Authorship::on_initialize(block_number);
+            assert!(ImOnline::is_online(AURA_AUTHOR_ID));
+        });
+    }
+
+    #[test]
+    fn authorship_notifies_staking() {
+        new_test_ext().execute_with(|| {
+            let block_number = 1;
+            let slot = Slot::from(SLOT_ID);
+            let pre_digest = Digest {
+                logs: vec![DigestItem::PreRuntime(AURA_ENGINE_ID, slot.encode())],
+            };
+            System::reset_events();
+            System::initialize(&block_number, &Default::default(), &pre_digest);
+
+            // Before authoring a block, no points have been given in the active era
+            assert!(
+                Staking::eras_reward_points(Staking::active_era().expect("No active era").index)
+                    .total
+                    == 0
+            );
+
+            // Pretend we author a block
+            Authorship::on_initialize(block_number);
+
+            // Authoring a block notifies Staking, which results in a positive points balance
+            assert!(
+                Staking::eras_reward_points(Staking::active_era().expect("No active era").index)
+                    .total
+                    > 0
+            );
+        });
+    }
+
+    #[test]
+    fn imonline_notifies_offences() {
+        new_test_ext().execute_with(|| {
+            let session = Session::current_index();
+            let offender_account = AccountId32::new(
+                testsfixtures::SAMPLE_USERS[AURA_AUTHOR_ID as usize]
+                    .raw_account
+                    .into(),
+            );
+
+            // Check that no previous offences were reported
+            let mut found = pallet_offences::ConcurrentReportsIndex::<Runtime>::get(
+                b"im-online:offlin",
+                session.encode(),
+            )
+            .into_iter()
+            .any(|offender| {
+                pallet_offences::Reports::<Runtime>::get(offender)
+                    .expect("Offence not found")
+                    .offender
+                    .0
+                    == offender_account
+            });
+            assert!(!found);
```

Comment:
I guess here introduce a function like `offenders()` that return an iterators of `AcccountId32` or a more specific `is_offender(AccointId)` function will make this test more readable and simpler to understand.

Reply:
Good idea, done!